zaza/openstack/charm_tests/ovn/tests.py

```diff
@@ -71,3 +76,231 @@ def setUpClass(cls):
+        self._resume_units('ovn-chassis')
+
+        # And we should be off to the races
+
+        self.one_time_init_done = True
+
+    def _add_neutron_api_plugin_ovn_subordinate_relation(self):
+        try:
+            logging.info('Adding relation neutron-api-plugin-ovn '
+                         '-> neutron-api')
+            zaza.model.add_relation(
+                'neutron-api-plugin-ovn', 'neutron-plugin',
+                'neutron-api:neutron-plugin-api-subordinate')
+            zaza.model.wait_for_agent_status()
+            zaza.model.wait_for_application_states(
+                states=self.test_config.get('target_deploy_status', {}))
+        except juju.errors.JujuAPIError:
+            # we were not able to add the relation, let's make sure it's
+            # because it's already there
+            assert (zaza.model.get_relation_id(
+                'neutron-api-plugin-ovn', 'neutron-api',
+                remote_interface_name='neutron-plugin-api-subordinate')
+                is not None), 'Unable to add relation required for test'
+            logging.info('--> On the other hand, did not need to add the '
+                         'relation as it was already there.')
+
+    def _configure_neutron_api(self):
+        """Set configuration option `manage-neutron-plugin-legacy-mode`."""
+        logging.info('Configuring `manage-neutron-plugin-legacy-mode` for '
+                     'neutron-api...')
+        n_api_config = {
+            'manage-neutron-plugin-legacy-mode': False,
+        }
+        with self.config_change(
+                n_api_config, n_api_config, 'neutron-api'):
+            logging.info('done')
+
+    def _run_offline_neutron_morph_db_action(self):
+        logging.info('Running the optional `offline-neutron-morph-db` action '
+                     'on neutron-api-plugin-ovn/leader')
+        generic_utils.assertActionRanOK(
+            zaza.model.run_action_on_leader(
+                'neutron-api-plugin-ovn',
+                'offline-neutron-morph-db',
+                action_params={
+                    'i-really-mean-it': True},
+                raise_on_failure=True,
+            )
+        )
+
+    def _run_migrate_ovn_db_action(self):
+        logging.info('Running `migrate-ovn-db` action on '
+                     'neutron-api-plugin-ovn/leader')
+        generic_utils.assertActionRanOK(
+            zaza.model.run_action_on_leader(
+                'neutron-api-plugin-ovn',
+                'migrate-ovn-db',
+                action_params={
+                    'i-really-mean-it': True},
+                raise_on_failure=True,
+            )
+        )
+
+    # Charm readiness is no guarantee for API being ready to serve requests.
+    # https://bugs.launchpad.net/charm-neutron-api/+bug/1854518
+    @tenacity.retry(wait=tenacity.wait_exponential(min=5, max=60),
+                    reraise=True, stop=tenacity.stop_after_attempt(3))
```

Comment:
Do we really want to re-run the action multiple times? Shouldn't we block and wait on some other indicator?

Reply:
The action is idempotent so it is safe to run multiple times. To avoid the tenacity we would have to fix the issue we have in all our charms where when the charm is ready the API/service it deploys often is not, I would love for us to fix that but I think that is out of scope for this work.

Reply:
Glad we have a bug for neutron-api that is a bug. But agree we can work around it here.